Detail function of each file in component documentation (#2825)
alphagov/govuk-frontend: docs/contributing/coding-standards/components.md

When creating your component, you should create the following files in the component’s folder:

> - `README.md` - Light documentation pointing to installation instructions and component documentation on <https://design-system.service.gov.uk/>

Reviewer: ```suggestion
- `README.md` - Concise documentation with links to the installation instructions and component documentation on <https://design-system.service.gov.uk/>
```

> - `_[component-name].scss` - A SCSS file to generate the styles just for the component, importing the variables, functions and mixins from `../../base` and the styles of the component (`./index.scss`).

Reviewer 2: ```suggestion
- `_[component-name].scss` - An SCSS file to generate the styles just for the component, importing the variables, functions and mixins from `../../base` and the styles of the component (`./index.scss`).
```

> - `_[component-name].scss` - An SCSS file to generate the styles just for the component, importing the variables, functions and mixins from `../../base` and the styles of the component (`./index.scss`).

Reviewer: ```suggestion
- `_[component-name].scss` - An SCSS file to generate the styles for the component, import the variables, functions and mixins from `../../base` and the styles of the component (`./index.scss`).
```

Author: I think this edit loses the meaning I was trying to express here, to help distinguish this file from `_index.scss`. This file is used so that people using the `govuk-frontend` package can [import the styles of a single component on its own](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#import-an-individual-component-s-css-using-a-single-import). 

Re-reading it, it feels like it's not really relevant how it does it (the part about importing the variables,...) as a developer will see that pretty clearly when they open the file, and there may be other parts getting added later on that'd require us to update that doc.  Maybe just that it delegates the actual styling to the `_index.scss` file to give people a heads up that there's not much going on in that file.

Maybe something like:
> An SCSS file to generate the styles for this component only. It delegates the CSS generation to the `_index.scss` file (see right after).

Reviewer: Ah yes, that is very clear. I don't think the bracket info at the end is needed but I'll amemd the suggested edit to your text.

> - `_index.scss` - The actual styles for the component, that can be imported either on their own by `[component-name].scss` or alongside other components in `components/_all.scss` (or by the consuming application)

Reviewer: ```suggestion
- `_index.scss` - The actual styles for the component that you can import in 2 ways:

-  on their own using `[component-name].scss`
- alongside other components in `components/_all.scss` (or with the consuming application)

```

Author: ```suggestion
- `_index.scss` - The actual styles for the component that you can import in 2 ways -  [on their own by `[component-name].scss`](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#import-specific-parts-of-the-css) or [alongside other components in `components/_all.scss`](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#import-specific-parts-of-the-css)
```

Reviewer: @romaricpascal this looks good, thank you.

> - `[component-name].yaml` - A list of the options expected by the component, used for listing in the documentation 

Reviewer 2: This file also includes examples of the component with specific options. These are used to generate examples in the documentation, as well as to run tests on.

Author: Thanks for clarifying that, I hadn't noticed it. Is that file also listing the sets of options used for the fixtures that Ollie was mentionning this morning?

Reviewer 2: Yup, those are used to generate the fixtures during certain Gulp tasks: https://github.com/alphagov/govuk-frontend/blob/54272096c11bf73c6e1fa5dd992e01c5e42a8a93/tasks/gulp/copy-to-destination.js#L56

Author: Cheers, I'll add that to the descrption as well, then.

> - `[component-name].yaml` - Lists the options expected by the component, used for display in the documentation. It also the option samples used to generate the examples in the documentation, as well as the [fixtures for testing alternative implementations of the design system](https://frontend.design-system.service.gov.uk/testing-your-html/).

Reviewer 2: ```suggestion
- `[component-name].yaml` - Lists the options expected by the component, used for display in the documentation. It also contains the option samples used to generate the examples in the documentation, as well as the [fixtures for testing alternative implementations of the design system](https://frontend.design-system.service.gov.uk/testing-your-html/).
```

> - `[component-name].yaml` - Lists the options expected by the component, used for display in the documentation. It also contains the option samples used to generate the examples in the documentation, as well as the [fixtures for testing alternative implementations of the design system](https://frontend.design-system.service.gov.uk/testing-your-html/).

Reviewer: ```suggestion
- `[component-name].yaml` - Lists the component's Nunjucks macro options, which are added to the component's documentation. It also contains the option samples used to generate the examples in the documentation, as well as the [fixtures for testing alternative implementations of the design system](https://frontend.design-system.service.gov.uk/testing-your-html/).
```

Reviewer 2: While it's correct that these are the Nunjucks macro options, we could possibly be more generic here, for the theoretical case where we ditch Nunjucks in favour of something else, these options would probably still be needed for whatever else we decided to use.

Maybe something like "Lists the parameters that can be passed to the component" (or less of a mouthful!).

Reviewer: OK, sounds sensible. I like your new sentence, but do you think it would be worth adding the Nunjucks macro as an example. Maybe like - "Lists the parameters (for example, Nunjucks macros) that are passed to the component." ? This sentence format also assumes that they definitely are passed - is that right or are they a possible, rather than a definite, action?

Reviewer: Yes, that's the one. I'll edit the suggestion to your description.

Reviewer 2: I've tweaked my wording - I was conflating the Design System website (which does use the options in its documentation), with this repo specifically.

- `macro.njk` - The main entry point for rendering the component, provides a `govuk[ComponentName](params)` macro, delegating render to the `template.njk` file
- `template.njk` - The template used for rendering the component using any `params` provided to the macro

> - `template.test.js` - Tests to ensure the component renders as intended with its various option

Reviewer 3: ```suggestion
- `template.test.js` - Tests to ensure the component renders as intended with its various options
```

> - `macro.njk` - The main entry point for rendering the component, provides a `govuk[ComponentName](params)` macro, delegating render to the `template.njk` file

Reviewer: ```suggestion
- `macro.njk` - The main entry point for rendering the component. It provides a `govuk[ComponentName](params)` macro, delegating render to the `template.njk` file
```

If your component uses JavaScript, you must also create the following files in the component’s folder:

- `[component-name].mjs` - A JavaScript module with the implementation of any behaviour needed by the component. See the [JavaScript documentation]('./js.md#skeleton) for a skeleton and more details on that file's structure

> - `[component-name].unit.test.mjs` - Unit tests to verify any lower-level logic used by the component individually

Reviewer: 'individually' is an orphan in this sentence. Is it still accurate to say 'Does individual unit tests to verify...' ?

Reviewer 2: Could maybe reword as "Unit tests to verify any component-specific lower-level logic"?

Basically, these are testing behaviour used internally by this component only.

Author: @domoscargin That sounds much better than the way I phrased it initially 😄

Reviewer: Yup, that's fab.

> - `[component-name].unit.test.mjs` - Unit tests to verify any lower-level logic used by the component individually

Reviewer: ```suggestion
- `[component-name].unit.test.mjs` - Unit tests to verify any component-specific lower-level logic.
```
